Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 namespace WebCore {
 
 VideoMediaSampleRenderer::VideoMediaSampleRenderer(WebSampleBufferVideoRendering *renderer)
-    : m_renderer(renderer)
+    : m_workQueue(WorkQueue::create("VideoMediaSampleRenderer Queue"_s))
+    , m_renderer(renderer)
 {
 }
 
 VideoMediaSampleRenderer::~VideoMediaSampleRenderer()
 {
-    [m_renderer flush];
-    [m_renderer stopRequestingMediaData];
-    if (m_decompressionSession)
-        m_decompressionSession->invalidate();
+    m_workQueue->dispatchSync([&] {
+        setTimebase(nullptr);
+
+        [m_renderer flush];
+        [m_renderer stopRequestingMediaData];
+        m_renderer = nil;
+
+        if (m_decompressionSession) {
+            m_decompressionSession->invalidate();
+            m_decompressionSession = nullptr;
+        }
+
+        m_hasDecompressionSession = false;
+    });
 }
 
 bool VideoMediaSampleRenderer::isReadyForMoreMediaData() const
 {
-    return (!m_decompressionSession || m_decompressionSession->isReadyForMoreMediaData()) && [m_renderer isReadyForMoreMediaData];
+    bool ready = false;
+    m_workQueue->dispatchSync([&] {
```

Comment:
we should avoid dispatchSync if possible. decompressionSession was made to be thread-safe earlier, that should continue.

Reply:
isReadyForMoreMediaData() is definitely not thread-safe.